src/main/java/org/mastodon/mamut/treesimilarity/ZhangUnorderedTreeEditDistance.java

```diff
@@ -0,0 +1,565 @@
+	{
+		double distance = forestDistances[ subtrees1.indexOf( forest1 ) ][ subtrees2.indexOf( forest2 ) ];
+		if ( distance != -1 )
+			return distance;
+		else
+		{
+			if ( !forest1.isLeaf() && forest2.isLeaf() )
+			{
+				forestDistances[ subtrees1.indexOf( forest1 ) ][ subtrees2.indexOf( forest2 ) ] = forestDeleteCosts.get( forest1 );
+				return forestDeleteCosts.get( forest1 );
+			}
+
+			if ( forest1.isLeaf() && !forest2.isLeaf() )
+			{
+				forestDistances[ subtrees1.indexOf( forest1 ) ][ subtrees2.indexOf( forest2 ) ] = forestInsertCosts.get( forest2 );
+				return forestInsertCosts.get( forest2 );
+			}
+
+			if ( !forest2.isLeaf() && !forest1.isLeaf() )
+			{
+				return getMinForestChangeCosts( forest1, forest2 );
+			}
+			throw new IllegalArgumentException( "The given trees are both leaves and thus they are both not forests." );
+		}
+	}
+
+	private double getMinTreeChangeCosts( Tree< T > tree1, Tree< T > tree2, @Nullable BiFunction< T, T, Double > costFunction )
+	{
+		double insertCosts = treeInsertCosts.get( tree2 );
+		List< Double > distances = new ArrayList<>();
+		if ( !tree2.isLeaf() )
+		{
+			for ( Tree< T > child : tree2.getChildren() )
+				distances.add( distanceTree( tree1, child, costFunction ) - treeInsertCosts.get( child ) );
+			insertCosts += Collections.min( distances );
+		}
+		double deleteCosts = treeDeleteCosts.get( tree1 );
+		distances = new ArrayList<>();
+		if ( !tree1.isLeaf() )
+		{
+			for ( Tree< T > child : tree1.getChildren() )
+				distances.add( distanceTree( child, tree2, costFunction ) - treeDeleteCosts.get( child ) );
+			deleteCosts += Collections.min( distances );
+		}
+		double changeCosts = distanceForest( tree1, tree2 );
+		if ( costFunction != null )
+			changeCosts += costTreeToTree.get( Pair.of( tree1, tree2 ) );
+
+		if ( tree1.isLeaf() || tree2.isLeaf() )
+		{
+			treeDistances[ subtrees1.indexOf( tree1 ) ][ subtrees2.indexOf( tree2 ) ] = changeCosts;
+			return changeCosts;
+		}
+		else
+		{
+			double minCost = min( insertCosts, deleteCosts, changeCosts );
+			treeDistances[ subtrees1.indexOf( tree1 ) ][ subtrees2.indexOf( tree2 ) ] = minCost;
+			return minCost;
+		}
+	}
+
+	private double getMinForestChangeCosts( Tree< T > forest1, Tree< T > forest2 )
+	{
+		double insertCosts = forestInsertCosts.get( forest2 );
+		List< Double > distances = new ArrayList<>();
+		if ( !forest2.isLeaf() )
+		{
+			for ( Tree< T > child : forest2.getChildren() )
+				distances.add( distanceForest( forest1, child ) - forestInsertCosts.get( child ) );
+			insertCosts += Collections.min( distances );
+		}
+		double deleteCosts = forestDeleteCosts.get( forest1 );
+		distances = new ArrayList<>();
+		if ( !forest1.isLeaf() )
+		{
+			for ( Tree< T > child : forest1.getChildren() )
+				distances.add( distanceForest( child, forest2 ) - forestDeleteCosts.get( child ) );
+			deleteCosts += Collections.min( distances );
+		}
```

Comment:
I don't see why this insert and delete costs are computed here. It makes no sense to me. This code can be removed with no consequences by doind the following.
```suggestion
		double insertCosts = Double.POSITIVE_INFINITY;
		double deleteCosts = Double.POSITIVE_INFINITY;
```
PLEASE CHECK: Is you translated this part of the python code correctly!

We need to improve the unit tests if this code is actually required, than it should be tested!

Reply:
Hm. Is this comment still applicable?

These lines are covered by multiple tests and the `insertCosts` and `deleteCosts` cannot be replaced by `Double.POSITIVE_INFINITY`, since this would let multiple tests fail.

Reply:
Discussed with @maarzt 
This comment is outdated.